torch/nn/modules/module.py

```diff
@@ -1181,7 +1181,10 @@ def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
 
         Args:
             state_dict (dict): a dict containing parameters and
-                persistent buffers.
+                persistent buffers. To load a dict from a DataParallel or
+                DistributedDataParallel module, need to first apply
```

Comment:
Is it possible to support this without changing `nn.Module` implementation? Otherwise, we are leaking a implementation detail of derived class into the base class. 

cc @gchanan @albanD

Reply:
Indeed, we either want to make this a public function or not document it. But not both.

Reply:
The issue we want to address here is loading a DP/DDP model from a non-DP/DDP model, so this util function is prepared for non-DP/DDP model users.

Reply:
> so this util function is prepared for non-DP/DDP model users.

Well not really. Because to get into this situation, you have to be using a DP/DDP. Or actually know that you're loading a model based on DP/DDP. So this is only for users that are aware that they use these modules.

Other options I would consider:
- Make DP and DDP strip this "module" layer by default when saving. And add it back automatically when loading.
- Document this explicitly in the DP/DDP to tell users they should serialize the .module or use this utils function
- Prevent saving DP/DDP directly and redirect the users to save the .module instead

What do you think?

Reply:
> Make DP and DDP strip this "module" layer by default when saving. And add it back automatically when loading.
I don't think this will work:
1) Some keys in DDP state_dict does not necessarily have the "module." prefix, so we cannot blindly add the prefix back for every key when loading.
2) If we change the loading logic in DDP, the DDP won't be compatible with any state_dict saved by previous DDP models.

> Document this explicitly in the DP/DDP to tell users they should serialize the .module or use this utils function
I think we can do this. You made a valid point that the user must be already aware that the saved model is a DDP model, so only adding the documentation in DDP should be fine.

> Prevent saving DP/DDP directly and redirect the users to save the .module instead
Again, this will hurt backward-compatibility if we prevent saving DP/DDP directly. We can add a soft suggestion in DDP documentation to courage users to save .module instead.

What do you think?

Reply:
> Some keys in DDP state_dict does not necessarily have the "module." prefix, so we cannot blindly add the prefix back for every key when loading.

Ho really? I thought that all the user modules are there and nothing else is around? 
And I don't think anyone in their right mind would subclass DP or DDP :p 

> If we change the loading logic in DDP, the DDP won't be compatible with any state_dict saved by previous DDP models.

We save versions in the serialization metadata just for that.
While loading, you would check the version, if it is old, load as-is, otherwise, implicitly append the `.module` before loading.

> I think we can do this. You made a valid point that the user must be already aware that the saved model is a DDP model, so only adding the documentation in DDP should be fine.

I think this is the right thing to do.
The only caveat is that there are so many warnings on these classes already, I am not sure users read them actually. Hence my more aggressive proposal to forbid this altogether. But this would be badly BC-breaking indeed. Note though that we could go through a deprecation cycle for this and then remove. @mrshenli do you think that is plausible? Or too many people rely on this?

Reply:
See this line:
https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L526

The key in `state_dict._metadata` does not necessarily have a "module." prefix, so implicitly appending a prefix when loading can be tricky.

I moved the notes from `nn.Module` to DDP to address this documentation scope issue. Even if we want to take the more aggressive approach that forbids DDP from saving a state dict directly, I think we should do it in a separate PR.

Reply:
I agree that this should be a separate issue if we want to do it.